source/common/http/codes.cc

```diff
@@ -18,111 +18,139 @@
+    rc_stat_names_[i] = nullptr;
+  }
+
+  // Pre-allocate response codes 200, 404, and 503, as those seem quite likely.
+  // We don't pre-allocate all the HTTP codes because the first 127 allocations
+  // are likely to be encoded in one byte, and we would rather spend those on
+  // common components of stat-names that appear frequently.
+  upstreamRqStatName(Code::OK);
+  upstreamRqStatName(Code::NotFound);
+  upstreamRqStatName(Code::ServiceUnavailable);
+}
+
+void CodeStatsImpl::incCounter(Stats::Scope& scope,
+                               const std::vector<Stats::StatName>& names) const {
+  Stats::SymbolTable::StoragePtr stat_name_storage = symbol_table_.join(names);
+  scope.counterFromStatName(Stats::StatName(stat_name_storage.get())).inc();
+}
+
+void CodeStatsImpl::incCounter(Stats::Scope& scope, Stats::StatName a, Stats::StatName b) const {
+  Stats::SymbolTable::StoragePtr stat_name_storage = symbol_table_.join({a, b});
+  scope.counterFromStatName(Stats::StatName(stat_name_storage.get())).inc();
+}
+
+void CodeStatsImpl::recordHistogram(Stats::Scope& scope, const std::vector<Stats::StatName>& names,
+                                    uint64_t count) const {
+  Stats::SymbolTable::StoragePtr stat_name_storage = symbol_table_.join(names);
+  scope.histogramFromStatName(Stats::StatName(stat_name_storage.get())).recordValue(count);
+}
+
+void CodeStatsImpl::chargeBasicResponseStat(Stats::Scope& scope, Stats::StatName prefix,
                                             Code response_code) const {
+  ASSERT(&symbol_table_ == &scope.symbolTable());
+
   // Build a dynamic stat for the response code and increment it.
-  scope.counter(absl::StrCat(prefix, upstream_rq_completed_)).inc();
-  scope
-      .counter(absl::StrCat(prefix, upstream_rq_,
-                            CodeUtility::groupStringForResponseCode(response_code)))
-      .inc();
-  scope.counter(absl::StrCat(prefix, upstream_rq_, enumToInt(response_code))).inc();
+  incCounter(scope, prefix, upstream_rq_completed_);
+  Stats::StatName rq_group = upstreamRqGroup(response_code);
+  if (!rq_group.empty()) {
+    incCounter(scope, prefix, rq_group);
+  }
+  incCounter(scope, prefix, upstreamRqStatName(response_code));
 }
 
 void CodeStatsImpl::chargeResponseStat(const ResponseStatInfo& info) const {
-  const uint64_t response_code = info.response_status_code_;
-  chargeBasicResponseStat(info.cluster_scope_, info.prefix_, static_cast<Code>(response_code));
+  Code code = static_cast<Code>(info.response_status_code_);
+
+  ASSERT(&info.cluster_scope_.symbolTable() == &symbol_table_);
+  chargeBasicResponseStat(info.cluster_scope_, info.prefix_, code);
+
+  Stats::StatName rq_group = upstreamRqGroup(code);
+  Stats::StatName rq_code = upstreamRqStatName(code);
 
-  std::string group_string =
-      CodeUtility::groupStringForResponseCode(static_cast<Code>(response_code));
+  auto write_category = [this, rq_group, rq_code, &info](Stats::StatName category) {
```

Comment:
perf nit: This could just be a member function and avoid the lambda creation?

Reply:
yes; done.